File: Haskell/problem016.hs
{-
Problem 16 - Project Euler
http://projecteuler.net/index.php?section=problems&id=16
-}
import Data.Char

-- powerPos :: Int -> Int -> Int
powerPos n p | p > 0 = n * powerPos n (p - 1)
powerPos _ 0 = 1
powerPos _ _ = 0

problem016 = sum . map (\a -> ord a - 48) . show $ powerPos 2 1000

main = putStrLn $ "problem016: answer " ++ show problem016

File: TypeScript/problem016.ts
/*
 * #16 Power Digit Sum - Project Euler
 * http://projecteuler.net/problem=16
 */
{
    function sumOfDigitsPower(base: number, exp: number): number {
        let digits = [1];

        for (let time = 0; time < exp; time++) {
            let carried = 0;

            const len = digits.length;
            for (let i = 0; i < len; i++) {
                const m = digits[i] * base + carried;
                carried = Math.floor(m / 10);
                digits[i] = m % 10;

                if (i == digits.length - 1 && carried > 0) {
                    digits.push(carried);
                }
            }
        }

        return digits.reduce((acc, x) => acc + x);
    }

    console.log(sumOfDigitsPower(2, 1000));
}
